File: app/src/main/java/com/pandatone/kumiwake/kumiwake/QuickKumiwakeResult.java
package com.pandatone.kumiwake.kumiwake;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import com.pandatone.kumiwake.R;
import com.pandatone.kumiwake.customize.CustomDialog;
import com.pandatone.kumiwake.sekigime.SekigimeResult;
import com.pandatone.kumiwake.sekigime.SelectTableType;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by atsushi_2 on 2016/05/10.
 */
public class QuickKumiwakeResult extends AppCompatActivity {

    ArrayList<String> memberArray, manArray, womanArray, groupArray, resultArray;
    static boolean even_fm_ratio;
    static boolean even_person_ratio;
    int groupNo;
    ArrayList<ArrayList<String>> arrayArray;
    RelativeLayout viewGroup;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.kumiwake_result);
        ButterKnife.bind(this);
        MobileAds.initialize(getApplicationContext(), "ca-app-pub-2315101868638564/8665451539");
        AdView mAdView = (AdView) findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice("BB707E3F7B5413908B2DD12063887489").build();
        mAdView.loadAd(adRequest);
        Intent i = getIntent();
        memberArray = i.getStringArrayListExtra("QuickModeMemberList");
        manArray = i.getStringArrayListExtra("QuickModeManList");
        womanArray = i.getStringArrayListExtra("QuickModeWomanList");
        groupArray = i.getStringArrayListExtra("QuickModeGroupList");
        even_fm_ratio = i.getBooleanExtra("EvenFMRatio", false);
        even_person_ratio = i.getBooleanExtra("EvenPersonRatio", false);
        groupNo = groupArray.size();
        viewGroup = (RelativeLayout) findViewById(R.id.result_view);
        viewGroup.setBackground(ContextCompat.getDrawable(this, R.drawable.quick_img));

        startMethod();

        if (!KumiwakeSelectMode.sekigime) {
            for (int v = 0; v < groupNo; v++) {
                resultArray = arrayArray.get(v);
                Collections.sort(resultArray, new KumiwakeComparator());
                addView(resultArray, v);
            }
        } else {
            for (int v = 0; v < groupNo; v++) {
                resultArray = arrayArray.get(v);
                Collections.shuffle(resultArray);
            }
            Intent intent = new Intent(this, SelectTableType.class);
            SekigimeResult.groupArray = groupArray;
            SekigimeResult.arrayArrayQuick = arrayArray;
            startActivity(intent);
            finish();
        }

        final ScrollView scrollView=(ScrollView)findViewById(R.id.kumiwake_scroll);
        scrollView.post(new Runnable() {
            public void run() {
                scrollView.fullScroll(ScrollView.FOCUS_UP);
                }});
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        for(int i=0;i<groupNo;i++){
            outState.putStringArrayList("ARRAY" + String.valueOf(i), arrayArray.get(i));}
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        arrayArray = new ArrayList<>(groupNo);
        for (int g = 0; g < groupNo; g++) {
            arrayArray.add(savedInstanceState.getStringArrayList("ARRAY" + String.valueOf(g)));
        }
    }


    public void startMethod() {
        arrayArray = new ArrayList<>(groupNo);

        for (int g = 0; g < groupNo; g++) {
            arrayArray.add(new ArrayList<String>());
        }

        if (even_fm_ratio == false) {
            kumiwakeAll();
        } else if (even_person_ratio == false) {
            kumiwakeFm(0);
        } else {
            kumiwakeFm(manArray.size() % groupNo);
        }
    }

    @OnClick(R.id.re_kumiwake)
    void onReKumiwake() {
        String title = getString(R.string.re_kumiwake_title);
        String message = getString(R.string.re_kumiwake_description) + getString(R.string.run_confirmation);
        final CustomDialog customDialog = new CustomDialog();
        customDialog.setTitle(title);
        customDialog.setMessage(message);
        CustomDialog.mPositiveBtnListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                                ScrollView scrollView = (ScrollView) findViewById(R.id.kumiwake_scroll);
                                scrollView.scrollTo(0, 0);
                                startMethod();
                                for (int i = 0; i < groupNo; i++) {
                                    resultArray = arrayArray.get(i);
                                    Collections.sort(resultArray, new KumiwakeComparator());
                                    addView(resultArray, i);
                                }
                                customDialog.dismiss();
                                Toast.makeText(getApplicationContext(), getText(R.string.re_kumiwake_finished), Toast.LENGTH_SHORT).show();
                            }
                        };
        customDialog.show(getFragmentManager(), "Btn");
    }

    @OnClick(R.id.go_sekigime)
    void onClicked() {
        for (int v = 0; v < groupNo; v++) {
            resultArray = arrayArray.get(v);
            Collections.shuffle(resultArray);
        }
        Intent intent = new Intent(this, SelectTableType.class);
        SekigimeResult.groupArray = groupArray;
        SekigimeResult.arrayArrayQuick = arrayArray;
        startActivity(intent);
    }

    @OnClick(R.id.go_home)
    void onClickedHome() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
    }


    public void kumiwakeAll() {
        Collections.shuffle(memberArray);
        int remainder;

        for (int i = 0; i < memberArray.size(); i++) {
            remainder = i % groupNo;
            arrayArray.get(remainder).add(memberArray.get(i));
        }
    }

    public void kumiwakeFm(int firstPos) {
        Collections.shuffle(manArray);
        Collections.shuffle(womanArray);
        int manRemainder, womanRemainder;

        for (int i = 0; i < manArray.size(); i++) {
            manRemainder = i % groupNo;
            arrayArray.get(manRemainder).add(manArray.get(i));
        }

        for (int i = 0; i < womanArray.size(); i++) {
            womanRemainder = (i + firstPos) % groupNo;
            arrayArray.get(womanRemainder).add(womanArray.get(i));
        }
    }


    public void addView(ArrayList<String> resultArray, int i) {
        TextView groupName;
        ListView arrayList;
        LinearLayout layout = (LinearLayout) findViewById(R.id.result_layout);
        View v = getLayoutInflater().inflate(R.layout.result_parts, null);
        if (i == 0) {
            layout.removeAllViews();
        }
        layout.addView(v);
        groupName = (TextView) v.findViewById(R.id.result_group);
        groupName.setText(getText(R.string.group) + " " + String.valueOf(i + 1));
        arrayList = (ListView) v.findViewById(R.id.result_member_listView);
        ArrayAdapter<String> adapter = new MemberArrayAdapter(this, R.layout.mini_row_member, resultArray, false);
        arrayList.setAdapter(adapter);
        setBackGround(v);
        QuickKumiwakeConfirmation.setRowHeight(arrayList, adapter);
    }

    public LinearLayout.LayoutParams setMargin(int leftDp, int topDp, int rightDp, int bottomDp) {
        float scale = getResources().getDisplayMetrics().density; //画面のdensityを指定。
        int left = (int) (leftDp * scale + 0.5f);
        int top = (int) (topDp * scale + 0.5f);
        int right = (int) (rightDp * scale + 0.5f);
        int bottom = (int) (bottomDp * scale + 0.5f);
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        layoutParams.setMargins(left, top, right, bottom);
        return layoutParams;
    }


    public void setBackGround(View v) {
        GradientDrawable drawable = new GradientDrawable();
        drawable.mutate();
        drawable.setShape(GradientDrawable.RECTANGLE);
        drawable.setCornerRadius(25);
        int R = 0, G = 0, B = 0;
        while (R < 200 && G < 200 && B < 200) {
            R = (int) ((Math.random() * 0.5 + 0.5) * 256);
            G = (int) ((Math.random() * 0.5 + 0.5) * 256);
            B = (int) ((Math.random() * 0.5 + 0.5) * 256);
        }

        drawable.setColor(Color.argb(150, R, G, B));
        v.setLayoutParams(setMargin(10, 12, 10, 0));
        v.setBackgroundDrawable(drawable);

    }
}


class KumiwakeComparator implements Comparator<String> {
    public int compare(String s1, String s2) {
        String s1_name = s1.substring(0, 5);
        String s2_name = s2.substring(0, 5);

        int value = s1_name.compareTo(s2_name);
        if (value == 0) {
            int s1_no = Integer.parseInt(s1.substring(5));
            int s2_no = Integer.parseInt(s2.substring(5));
            if (s1_no < s2_no) {
                value = -1;
            } else {
                value = 1;
            }
        }
        return value;
    }
}

File: app/src/main/java/com/pandatone/kumiwake/kumiwake/QuickKumiwakeResult.kt
package com.pandatone.kumiwake.kumiwake

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.app.ShareCompat
import androidx.core.content.ContextCompat
import butterknife.ButterKnife
import butterknife.OnClick
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.MobileAds
import com.pandatone.kumiwake.MainActivity
import com.pandatone.kumiwake.QuickModeKeys
import com.pandatone.kumiwake.R
import com.pandatone.kumiwake.StatusHolder
import com.pandatone.kumiwake.sekigime.SekigimeResult
import com.pandatone.kumiwake.sekigime.SelectTableType
import com.pandatone.kumiwake.ui.dialogs.DialogWarehouse
import java.util.*

/**
 * Created by atsushi_2 on 2016/05/10.
 */
class QuickKumiwakeResult : AppCompatActivity() {

    private lateinit var memberArray: ArrayList<String>
    private lateinit var manArray: ArrayList<String>
    private lateinit var womanArray: ArrayList<String>
    private lateinit var groupArray: ArrayList<String>
    private lateinit var resultArray: ArrayList<String>
    private var groupNo: Int = 0
    private var evenFmRatio: Boolean = false
    private lateinit var arrayArray: ArrayList<ArrayList<String>>
    private lateinit var viewGroup: ConstraintLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.kumiwake_result)
        ButterKnife.bind(this)
        MobileAds.initialize(applicationContext, "ca-app-pub-2315101868638564/8665451539")
        val mAdView = findViewById<View>(R.id.adView) as AdView
        val adRequest = AdRequest.Builder()
                .addTestDevice("BB707E3F7B5413908B2DD12063887489").build()
        mAdView.loadAd(adRequest)
        val i = intent.also {
            memberArray = it.getStringArrayListExtra(QuickModeKeys.MEMBER_LIST.key)
            manArray = it.getStringArrayListExtra(QuickModeKeys.MAN_LIST.key)
            womanArray = it.getStringArrayListExtra(QuickModeKeys.WOMAN_LIST.key)
            groupArray = it.getStringArrayListExtra(QuickModeKeys.GROUP_LIST.key)
            evenFmRatio = it.getBooleanExtra(QuickModeKeys.EVEN_FM_RATIO.key, false)
        }
        groupNo = groupArray.size

        viewGroup = findViewById<View>(R.id.result_view) as ConstraintLayout
        viewGroup.background = ContextCompat.getDrawable(this, R.drawable.quick_img)

        startMethod()

        if (!StatusHolder.sekigime) {
            for (v in 0 until groupNo) {
                resultArray = arrayArray[v]
                Collections.sort(resultArray, KumiwakeComparator())
                addView(resultArray, v)
            }
        } else {
            for (v in 0 until groupNo) {
                resultArray = arrayArray[v]
                resultArray.shuffle()
            }
            val intent = Intent(this, SelectTableType::class.java)
            SekigimeResult.groupArray = groupArray
            SekigimeResult.arrayArrayQuick = arrayArray
            startActivity(intent)
            finish()
        }

        val scrollView = findViewById<View>(R.id.kumiwake_scroll) as ScrollView
        scrollView.post { scrollView.fullScroll(ScrollView.FOCUS_UP) }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        for (i in 0 until groupNo) {
            outState.putStringArrayList("ARRAY$i", arrayArray[i])
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)

        arrayArray = ArrayList(groupNo)
        for (g in 0 until groupNo) {
            savedInstanceState.getStringArrayList("ARRAY$g")?.let { arrayArray.add(it) }
        }
    }


    private fun startMethod() {
        arrayArray = ArrayList(groupNo)

        for (g in 0 until groupNo) {
            arrayArray.add(ArrayList())
        }

        if (evenFmRatio) {
            kumiwakeFm(manArray.size % groupNo)
        } else {
            kumiwakeAll()
        }
    }

    @OnClick(R.id.re_kumiwake)
    internal fun onReKumiwake() {
        val title = getString(R.string.re_kumiwake_title)
        val message = getString(R.string.re_kumiwake_description) + getString(R.string.run_confirmation)
        DialogWarehouse(supportFragmentManager).decisionDialog(title,message, this::reKumiwake)
    }

    private fun reKumiwake(){
        val scrollView = findViewById<View>(R.id.kumiwake_scroll) as ScrollView
        scrollView.scrollTo(0, 0)
        startMethod()
        for (i in 0 until groupNo) {
            resultArray = arrayArray[i]
            Collections.sort(resultArray, KumiwakeComparator())
            addView(resultArray, i)
        }
        Toast.makeText(applicationContext, getText(R.string.re_kumiwake_finished), Toast.LENGTH_SHORT).show()
    }

    @OnClick(R.id.share_result)
    internal fun shareResult() {
        share()
    }

    @OnClick(R.id.go_sekigime)
    internal fun onClicked() {
        for (v in 0 until groupNo) {
            resultArray = arrayArray[v]
            resultArray.shuffle()
        }
        val intent = Intent(this, SelectTableType::class.java)
        SekigimeResult.groupArray = groupArray
        SekigimeResult.arrayArrayQuick = arrayArray
        startActivity(intent)
    }

    @OnClick(R.id.go_home)
    internal fun onClickedHome() {
        val intent = Intent(this, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(intent)
    }


    private fun kumiwakeAll() {
        memberArray.shuffle()
        var targetGroupNo: Int

        for (i in memberArray.indices) {
            //メンバーを追加するグループを回す(0-groupNo)
            targetGroupNo = i % groupNo
            arrayArray[targetGroupNo].add(memberArray[i])
        }
    }

    private fun kumiwakeFm(firstPos: Int) {
        manArray.shuffle()
        womanArray.shuffle()
        var manTargetGroupNo: Int
        var womanTargetGroupNo: Int

        for (i in manArray.indices) {
            //メンバーを追加するグループを回す(0-groupNo)
            manTargetGroupNo = i % groupNo
            arrayArray[manTargetGroupNo].add(manArray[i])
        }

        for (i in womanArray.indices) {
            //メンバーを追加するグループを回す(0-groupNo)
            womanTargetGroupNo = (i + firstPos) % groupNo
            arrayArray[womanTargetGroupNo].add(womanArray[i])
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    /////////////////////////  View  ///////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////

    @SuppressLint("SetTextI18n", "InflateParams")
    fun addView(resultArray: ArrayList<String>, i: Int) {
        val groupName: TextView
        val arrayList: ListView
        val layout = findViewById<View>(R.id.result_layout) as LinearLayout
        val v = layoutInflater.inflate(R.layout.result_parts, null)
        if (i == 0) {
            layout.removeAllViews()
        }
        layout.addView(v)
        groupName = v.findViewById<View>(R.id.result_group) as TextView
        groupName.text = getText(R.string.group).toString() + " " + (i + 1).toString()
        arrayList = v.findViewById<View>(R.id.result_member_listView) as ListView
        val adapter = MemberArrayAdapter(this, R.layout.mini_row_member, resultArray, false)
        arrayList.adapter = adapter
        setBackGround(v)
        QuickKumiwakeConfirmation.setRowHeight(arrayList, adapter)
    }

    private fun setMargin(leftDp: Int, topDp: Int, rightDp: Int, bottomDp: Int): LinearLayout.LayoutParams {
        val scale = resources.displayMetrics.density //画面のdensityを指定。
        val left = (leftDp * scale + 0.5f).toInt()
        val top = (topDp * scale + 0.5f).toInt()
        val right = (rightDp * scale + 0.5f).toInt()
        val bottom = (bottomDp * scale + 0.5f).toInt()
        val layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        layoutParams.setMargins(left, top, right, bottom)
        return layoutParams
    }


    private fun setBackGround(v: View) {
        val drawable = GradientDrawable()
        drawable.mutate()
        drawable.shape = GradientDrawable.RECTANGLE
        drawable.cornerRadius = 25f
        var R = 0
        var G = 0
        var B = 0
        while (R < 200 && G < 200 && B < 200) {
            R = ((Math.random() * 0.5 + 0.5) * 256).toInt()
            G = ((Math.random() * 0.5 + 0.5) * 256).toInt()
            B = ((Math.random() * 0.5 + 0.5) * 256).toInt()
        }

        drawable.setColor(Color.argb(150, R, G, B))
        v.layoutParams = setMargin(10, 12, 10, 0)
        v.background = drawable

    }

    private fun share() {
        val articleTitle = "～" + getString(R.string.kumiwake_result) + "～"
        val resultTxt = StringBuilder()
        for ((i, array) in arrayArray.withIndex()) {
            resultTxt.append("\n")
            resultTxt.append("《${groupArray[i]}》\n")

            for (member in array) {
                resultTxt.append("$member\n")
            }
        }

        val sharedText = "$articleTitle\n$resultTxt"

        // builderの生成　ShareCompat.IntentBuilder.from(Context context);
        val builder = ShareCompat.IntentBuilder.from(this)

        // アプリ一覧が表示されるDialogのタイトルの設定
        builder.setChooserTitle(R.string.choose_app)

        // シェアするタイトル
        builder.setSubject(articleTitle)

        // シェアするテキスト
        builder.setText(sharedText)

        // シェアするタイプ（他にもいっぱいあるよ）
        builder.setType("text/plain")

        // Shareアプリ一覧のDialogの表示
        builder.startChooser()

    }
}


internal class KumiwakeComparator : Comparator<String> {
    override fun compare(s1: String, s2: String): Int {
        val s1Name = s1.replace("[0-9]".toRegex(), "") //文字列から文字のみ抜き出し
        val s2Name = s2.replace("[0-9]".toRegex(), "")

        var value = s1Name.compareTo(s2Name)

        if (value == 0) {
            val s1No = Integer.parseInt(s1.replace("[^0-9]".toRegex(), ""))   //文字列から数値のみ抜き出し
            val s2No = Integer.parseInt(s2.replace("[^0-9]".toRegex(), ""))
            value = if (s1No < s2No) {
                -1
            } else {
                1
            }
        }


        return value
    }
}

